clients/drcachesim/tracer/output.cpp

```diff
@@ -1145,6 +1145,21 @@ exit_thread_io(void *drcontext)
 {
     per_thread_t *data = (per_thread_t *)drmgr_get_tls_field(drcontext, tls_idx);
 
+#ifdef LINUX
+    /**
+     * On Linux, the thread exit event may be invoked twice for the same thread
+     * if that thread is alive during a process fork, but doesn't call the fork
+     * itself.  The first time the event callback is executed from the fork child
+     * immediately after the fork, the second time it is executed during the
+     * regular thread exit.
+     * data->file could be already closed. Write file operation is fail
+     * and it is asserted.
+     */
+    if (dr_get_process_id() != dr_get_process_id_from_drcontext(drcontext)) {
```

Comment:
It would be good to add a test for this.  We have `tool.drcacheoff.fork` which runs `linux.fork` but that is single-threaded.  Could we change it to run `pthreads.pthreads_fork_FLAKY` and reproduce the assert and verify the fix?  I don't know what the `_FLAKY` is from: hopefully just non-deterministic output ordering and we can make the template looser than the plain expect file?

Reply:
`$BLD_DRI/bin64/drrun -t drcachesim -offline $BLD_DRI/suite/tests/bin/pthreads.pthreads_fork_FLAKY` reproduced the same assert.
```
ASSERT FAILURE: $BLD_DRI/clients/drcachesim/tracer/output.cpp:531: static_cast<size_t>(wrote) == res ()
child 3537435 exited non-zero: ff00
signalled: 0, signal: 0, exited: 1, exit: 255
all done
```

This issue is fixed with my patch but tool is still crashed. I've tried different tools and caught that opcode_count is crashed in the same way too but opcodes are ok.
I wrote 2 small tools: one is ok and the 2nd is crashed
OK
```
#include "dr_api.h"
#include "drmgr.h"
#include "drx.h"

static void
event_exit(void)
{
    drx_exit();
    drmgr_exit();
}

DR_EXPORT void
dr_client_main(client_id_t id, int argc, const char *argv[])
{
    if (!drmgr_init())
        DR_ASSERT(false);
    drx_init();

    dr_register_exit_event(event_exit);
}
```

FAIL
```
#include "dr_api.h"
#include "drmgr.h"
#include "drx.h"
#include "droption.h"

static void
event_exit(void)
{
    drx_exit();
    drmgr_exit();
}

DR_EXPORT void
dr_client_main(client_id_t id, int argc, const char *argv[])
{
    if (!droption_parser_t::parse_argv(DROPTION_SCOPE_CLIENT, argc, argv, NULL, NULL))
        DR_ASSERT(false);

    if (!drmgr_init())
        DR_ASSERT(false);
    drx_init();

    dr_register_exit_event(event_exit);
}
```

`$BLD_DRI/bin64/drrun -c $BLD_DRI/api/bin/libFAIL.so -- $BLD_DRI/suite/tests/bin/pthreads.pthreads_fork_FLAKY`
error output
```
child process
child process
<Application $BLD_DRI/suite/tests/bin/pthreads.pthreads_fork_FLAKY (3537560).  Tool internal crash at PC 0x00007f3a3964ffdb.  Please report this at your tool's issue tracker.  Program aborted.
Received SIGSEGV at pc 0x00007f3a3964ffdb in thread 3537560
Base: 0x00007f3a39dde000
Registers:eax=0x0000000000000001 ebx=0x00007f3a397f6d90 ecx=0x0000000000000008 edx=0x0000000000000001
        esi=0x0000000000000000 edi=0x00007f3a39bca178 esp=0x00007f37f600e450 ebp=0x00007f3a39bca000
        r8 =0x3b198bb19e0f8818 r9 =0x00007f3a39bca178 r10=0x00007f39f5de6000 r11=0x0000000000000202
        r12=0x00007f3a397f6cb0 r13=0x00007f3a397fa2e8 r14=0x0000000000000008 r15=0x00007f3a397f6ca0
        eflags=0x0000000000010246
version 9.0.19306, custom build
-no_dynamic_options -client_lib '$BLD_DRI/api/bin/libFAIL.so;0;' -client_lib64 '$BLD_DRI/api/bin/libFAIL.so;0;' -code_api -stack_size 56K -signa
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8>
child 3537560 exited non-zero: ff00
signalled: 0, signal: 0, exited: 1, exit: 255
<Application $BLD_DRI/suite/tests/bin/pthreads.pthreads_fork_FLAKY (3537556).  Tool internal crash at PC 0x00007f3a3964ffdb.  Please report this at your tool's issue tracker.  Program aborted.
Received SIGSEGV at pc 0x00007f3a3964ffdb in thread 3537556
Base: 0x00007f3a39dde000
Registers:eax=0x0000000000000001 ebx=0x00007f3a397f6d90 ecx=0x0000000000000008 edx=0x0000000000000001
        esi=0x0000000000000000 edi=0x00007f3a39bca178 esp=0x00007f37f5fb5450 ebp=0x00007f3a39bca000
        r8 =0x3b198bb19e0f8818 r9 =0x00007f3a39bca178 r10=0x00007f39f5de6000 r11=0x0000000000000202
        r12=0x00007f3a397f6cb0 r13=0x00007f3a397fa2e8 r14=0x0000000000000008 r15=0x00007f3a397f6ca0
        eflags=0x0000000000010246
version 9.0.19306, custom build
-no_dynamic_options -client_lib '$BLD_DRI/api/bin/libFAIL.so;0;' -client_lib64 '$BLD_DRI/api/bin/libFAIL.so;0;' -code_api -stack_size 56K -signa
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8
0x00007f3a39bca000 0x00007f3a39bbf0f8>
child 3537556 exited non-zero: ff00
signalled: 0, signal: 0, exited: 1, exit: 255
all done
---------------
child process
child process
all done
```

Looks like there is some another global issue with forks that is reproduces on pthread_fork test examples. Currently I didn't see something like this on our java workloads.
Could you look at that? Maybe you know what could be wrong here.
Thanks, Kirill

Reply:
The new issue is on x86. Could not reproduce tool crashes on AArch64.

Write issue is not reproduced with patch on both x86 and AARch64.

Reply:
Create separate issue for taht #5735